File: src/main/java/com/camackenzie/exvi/core/api/APIResult.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.camackenzie.exvi.core.api;

import com.google.gson.Gson;
import java.util.HashMap;

/**
 *
 * @author callum
 */
public class APIResult<T> {

    private static Gson gson = new Gson();

    private T body;
    private int statusCode;
    private HashMap<String, String> headers;

    public APIResult(int statusCode, T body, HashMap<String, String> headers) {
        this.statusCode = statusCode;
        this.body = body;
        this.headers = headers;
    }

    public APIResult(APIResult other, T newBody) {
        this.body = newBody;
        this.statusCode = other.getStatusCode();
        this.headers = other.getHeaders();
    }

    public APIResult<T> withJsonHeader() {
        this.headers.put("content-type", "application/json");
        return this;
    }

    public T getBody() {
        return this.body;
    }

    public void setBody(T body) {
        this.body = body;
    }

    public HashMap<String, String> getHeaders() {
        return this.headers;
    }

    public void setHeaders(HashMap<String, String> headers) {
        this.headers = headers;
    }

    public int getStatusCode() {
        return this.statusCode;
    }

    public void setStatusCode(int e) {
        this.statusCode = e;
    }

    public String toJson() {
        this.withJsonHeader();
        return gson.toJson(this);
    }

    public static <T> APIResult<T> jsonResult(int statusCode, T body) {
        return new APIResult(statusCode, body, new HashMap<>()).withJsonHeader();
    }

    public static <T> String jsonString(int statusCode, T body) {
        return APIResult.<T>jsonResult(statusCode, body).toJson();
    }

}

File: src/commonMain/kotlin/com/camackenzie/exvi/core/api/APIResult.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.camackenzie.exvi.core.api

import kotlinx.serialization.*
import kotlinx.serialization.json.*

/**
 *
 * @author callum
 */
@Serializable
class APIResult<T> {
    var body: T
    var statusCode: Int
    var headers: HashMap<String, String>

    constructor(statusCode: Int, body: T, headers: HashMap<String, String>) {
        this.statusCode = statusCode
        this.body = body
        this.headers = headers
    }

    constructor(other: APIResult<*>, newBody: T) {
        body = newBody
        statusCode = other.statusCode
        headers = other.headers
    }

    fun withJsonHeader(): APIResult<T> {
        headers["content-type"] = "application/json"
        return this
    }

    fun toJson(): String {
        withJsonHeader()
        return Json.encodeToString(this)
    }

    fun failed(): Boolean {
        return statusCode != 200
    }

    fun succeeded(): Boolean {
        return !failed()
    }

    companion object {
        fun <T> jsonResult(statusCode: Int, body: T): APIResult<T> {
            return APIResult(statusCode, body, HashMap<String, String>()).withJsonHeader()
        }

        fun <T> jsonString(statusCode: Int, body: T): String {
            return jsonResult(statusCode, body).toJson()
        }
    }
}

inline fun <reified T> APIResult<String>.decodeBody(): T {
    return Json.decodeFromString(this.body)
}